src/libraries/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.OpenSsl.cs

```diff
@@ -821,6 +814,42 @@ private static void BioWrite(SafeBioHandle bio, ReadOnlySpan<byte> buffer)
             return innerError;
         }
 
+        private static void SetSslCertificate(SafeSslContextHandle contextPtr, X509Certificate2 cert)
+        {
+            SafeEvpPKeyHandle? certKeyHandle = null;
+
+            using (RSAOpenSsl? rsa = (RSAOpenSsl?)cert.GetRSAPrivateKey())
+            {
+                if (rsa != null)
+                {
+                    certKeyHandle = rsa.DuplicateKeyHandle();
+                    Interop.Crypto.CheckValidOpenSslHandle(certKeyHandle);
```

Comment:
If this throws because certKeyHandle.IsInvalid is true, should certKeyHandle be disposed?

Reply:
good questions. This is essentially copy from 
https://github.com/dotnet/runtime/blob/9af8d302389305d2efaf1fd8db5c1dea0a246900/src/libraries/Common/src/System/Net/Security/Unix/SafeFreeSslCredentials.cs#L46-L93

Since this all comes from single object e.g. `X509Certificate2` I'm not sure if that is possible. 
cc: @bartonjs  

I'de be happy to add code to deal with it instead of depending on finalizer.

Reply:
good questions. This is essentially copy from 
https://github.com/dotnet/runtime/blob/9af8d302389305d2efaf1fd8db5c1dea0a246900/src/libraries/Common/src/System/Net/Security/Unix/SafeFreeSslCredentials.cs#L46-L93

Since this all comes from single object e.g. `X509Certificate2` I'm not sure if that is possible. 
cc: @bartonjs  

I'de be happy to add code to deal with it instead of depending on finalizer.

Reply:
The code has changed a bit in the years since.  RSAOpenSsl::DuplicateKeyHandle will never return an IsInvalid SafeHandle, it would have already disposed the object and thrown.

https://github.com/dotnet/runtime/blob/eef6c7f6cb7f73163313d349fbb796f37b4c3996/src/libraries/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.EvpPkey.cs#L21-L38

Reply:
I realized we don't need to do this for each session. Moved the handle logic to `SslStreamCertificateContext.Linux` so it is done once upfront.